<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.17"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>linearham: Main Page</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    extensions: ["tex2jax.js"],
    jax: ["input/TeX","output/HTML-CSS"],
});
</script>
<script type="text/javascript" async="async" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="style.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">linearham
   </div>
   <div id="projectbrief">A Phylo-HMM implementation for B cell receptor sequence analysis.</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.17 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="PageDoc"><div class="header">
  <div class="headertitle">
<div class="title">linearham Documentation</div>  </div>
</div><!--header-->
<div class="contents">
<div class="toc"><h3>Table of Contents</h3>
<ul><li class="level1"><a href="#sw_alignment">Smith-Waterman alignment information</a></li>
<li class="level1"><a href="#germline_padding">Germline padding</a></li>
<li class="level1"><a href="#vdj_germline">[VDJ]Germline classes</a></li>
<li class="level1"><a href="#state_space">HMM hidden state space</a></li>
<li class="level1"><a href="#transition_prob">HMM hidden state transition probability matrices</a></li>
<li class="level1"><a href="#emission_prob">HMM hidden state emission probability matrices</a></li>
<li class="level1"><a href="#xmsa">Phylo-HMM &quot;expanded&quot; multiple sequence alignment (xMSA)</a></li>
<li class="level1"><a href="#scaling">HMM scaling for numeric underflow</a></li>
</ul>
</div>
<div class="textblock"><p><code>linearham</code> is a phylogenetic hidden Markov model (phylo-HMM), which simultaneously models the VDJ recombination process (with an HMM) and sequence evolution (with a phylogenetic tree).</p>
<p>This document is the reference for understanding how <code>linearham</code> works. In the following sections, we provide an overview of the abstractions used in the <code>linearham</code> codebase. See the links above for detailed information about classes and methods. To fully understand how <code>linearham</code> works, please see the <code>linearham</code> tests, which provide examples describing these core concepts.</p>
<p>We use the term "site positions" to refer to positions in the aligned sequences and "germline positions" to positions in the germline gene.</p>
<h1><a class="anchor" id="sw_alignment"></a>
Smith-Waterman alignment information</h1>
<p>To allow for more tractable inference on the HMM, we utilize Smith-Waterman (S-W) alignment information between the clonal family sequences and the germline genes. We fix the relative positions of the germline genes to their S-W site positions. Of course, the S-W algorithm aligns two sequences at a time, but we use heuristics to determine the final alignment between all the relevant germline genes and the equal-length clonal sequences. For more information on these heuristics, see the <code>add_linearham_info()</code> function in <code>python/utils.py</code> of <code>partis</code> (<a href="https://github.com/psathyrella/partis/tree/dev">https://github.com/psathyrella/partis/tree/dev</a>).</p>
<p>The diagram shown below provides an example alignment between a single-sequence clonal family and a V/D/J gene. In the code, we have two important S-W information objects: <code>flexbounds_</code> and <code>relpos_</code>. <code>flexbounds_</code> is a map holding (<code>[vdj]_[lr]</code>, <code>pair&lt;int, int&gt;</code>) pairs, which represent the possible V/D/J starting/ending match positions. During inference, these starting/ending match position ranges are found by computing the <code>min</code>/<code>max</code> of the corresponding S-W starting/ending positions for all germline gene matches in a given gene type (i.e. V/D/J). In addition, the <code>[vd]_r</code> (<code>[dj]_l</code>) bounds are shifted to the left (right) by a certain amount to allow for more flexible naive sequence inference in the CDR3 region.</p>
<p>This alignment graphic is a simple example to demonstrate how these data structures work. Because the S-W match positions for a given germline gene are specified using Python 0-based right-exclusive range conventions, the <code>[vdj]_r</code> bounds in <code>flexbounds_</code> denote the range of site positions immediately <em>after</em> the last possible S-W match positions in a given gene type. In this example,</p>
<div class="fragment"><div class="line">flexbounds_ = {{<span class="stringliteral">&quot;v_l&quot;</span>, {0, 2}} , {<span class="stringliteral">&quot;v_r&quot;</span>, {4, 6}}  , {<span class="stringliteral">&quot;d_l&quot;</span>, {7, 8}},</div>
<div class="line">               {<span class="stringliteral">&quot;d_r&quot;</span>, {9, 10}}, {<span class="stringliteral">&quot;j_l&quot;</span>, {11, 12}}, {<span class="stringliteral">&quot;j_r&quot;</span>, {15, 15}}}</div>
</div><!-- fragment --><p>These ranges are marked in the diagram between vertical dashed lines. Note that the last V gene starting position is 2 and the first V gene ending position is 4, which means only positions 2 and 3 are guaranteed to be in the V gene. This logic will be important when we describe how the HMM hidden state space is constructed. <code>relpos_</code> is a map specifying the starting positions of the germline genes. In this example, <code>relpos_ = {{"V", 1}, {"D", 5}, {"J", 10}}</code>.</p>
<div class="image">
<img src="sw_alignment.jpg" alt=""/>
</div>
<h1><a class="anchor" id="germline_padding"></a>
Germline padding</h1>
<p>In the diagram above, the germline genes are not properly aligned to the clonal family sequence as the V (J) gene should align to the start (end) of the sequence. To account for this, we pad germline genes with N bases until the V (J) gene aligns to the start (end) of the clonal sequence. These padding bases represent fully ambiguous (i.e. any possible) germline bases. The padding transition/emission probability information is stored in the <code>NPadding</code> class. Note that while germline padding is handled by <code>linearham</code>, sequence padding is performed in <code>partis</code>.</p>
<h1><a class="anchor" id="vdj_germline"></a>
[VDJ]Germline classes</h1>
<p>We construct three germline gene classes, one for each type of germline gene (i.e. V/D/J). These classes <code>VGermline</code>, <code>DGermline</code>, and <code>JGermline</code> are inherited from three classes that store <code>partis</code> HMM germline parameter information (i.e. <code>Germline</code>, <code>NTInsertion</code>, and <code>NPadding</code>). The corresponding inheritance diagram is shown below.</p>
<div class="dotgraph">
<img src="dot_inline_dotgraph_1.png" alt="dot_inline_dotgraph_1.png" border="0" usemap="#dot_inline_dotgraph_1.map"/>
</div>
<p>Looking at this diagram, you can see that we account for non-templated insertions to the left of D and J germline genes. In the code, we parse the parameter files (in YAML format) and store these <code>[VDJ]Germline</code> objects in a common <code>GermlineGene</code> class, which is conceptually similar to a tagged union class.</p>
<h1><a class="anchor" id="state_space"></a>
HMM hidden state space</h1>
<p>Our hidden state space is similar to that of <a href="https://github.com/psathyrella/ham"><code>ham</code></a>. <code>ham</code> uses states of the form \((V, j)\), \((D, j)\), \((D, N)\), \((J, j)\), and \((J, N)\) for a V gene \(V\), D gene \(D\), J gene \(J\), germline position \(j\), and non-templated insertion base \(N \in \{N_A, N_C, N_G, N_T\}\). Again, non-templated insertion bases appear to the left of a germline gene and thus are associated with the gene to their right. In <code>linearham</code> (versus <code>ham</code>), we use the S-W alignment information to collapse the germline state space.</p>
<p>As described in the above S-W alignment diagram, positions 2 and 3 represent the sites "guaranteed" to be in a hidden V germline state according to S-W. In this "germline" region, we do not need to care about the different possibilities of hidden states beyond which V gene is entered because we know that the HMM must march along the gene until it exits that gene. Thus, the only possible hidden state in the V "germline" region is \(V\). It is helpful to think about the V "germline" region as a single site position in the HMM.</p>
<p>To make this more concrete, we provide another S-W alignment diagram below with more than one V/D/J germline gene. In this new example, the hidden V "germline" state can be either \(V_{01}\) or \(V_{99}\). The V-D "junction" region (i.e. sites 4 and 5) can have hidden states associated with V genes, non-templated insertions, and D genes. Specifically, the hidden state space consists of \((V_{01}, 3)\), \((V_{01}, 4)\), \((V_{99}, 3)\), \((V_{99}, 4)\), \((D_{01}, N_A)\), \((D_{01}, N_C)\), \((D_{01}, N_G)\), \((D_{01}, N_T)\), \((D_{01}, 0)\), \((D_{99}, N_A)\), \((D_{99}, N_C)\), \((D_{99}, N_G)\), \((D_{99}, N_T)\), \((D_{99}, 1)\), \((D_{99}, 2)\). In general, the hidden state space in the V-D "junction" region comprises the matched V germline states, the non-templated insertion states associated with D gene matches, and the matched D germline states from site position <code>flexbounds_["v_r"].first</code> to site position <code>flexbounds_["d_l"].second - 1</code>; the site position <code>flexbounds_["d_l"].second</code> is the last D gene match starting position so it is considered part of the D "germline" region (and not part of the V-D "junction" region). Note that this discussion about the V "germline" and V-D "junction" states applies to the D/J "germline" and D-J "junction" states as well.</p>
<div class="image">
<img src="sw_alignment_extra.jpg" alt=""/>
</div>
<p>This "germline" and "junction" state space decomposition allows us to reduce the space-time complexity of the forward algorithm, which is quadratic in the number of hidden states. By collapsing the "germline" state space using S-W alignment information, the forward algorithm scales approximately linearly in the number of <code>ham</code> hidden states (hence the name <code>linearham</code>).</p>
<h1><a class="anchor" id="transition_prob"></a>
HMM hidden state transition probability matrices</h1>
<p>Because the <code>linearham</code> HMM has different hidden state spaces for the "germline" and "junction" regions, we need to specify "germline"-to-"junction", "junction", and "junction"-to-"germline" hidden state transition probability matrices. To help illustrate how these matrices are structured, we revisit the S-W example alignment first discussed in <a class="el" href="index.html#sw_alignment">Smith-Waterman alignment information</a>. Specifically, we focus our attention on the V "germline" region, V-D "junction" region, and D "germline" region; the other "germline" and "junction" regions can be treated analogously.</p>
<p>The hidden state spaces in that first example for the V "germline" region, V-D "junction" region, and D "germline" region are \(\{V\}\), \(\{(V, 3)\), \((V, 4)\), \((D, N_A)\), \((D, N_C)\), \((D, N_G)\), \((D, N_T)\), \((D, 0)\), \((D, 1)\), \((D, 2)\}\), and \(\{D\}\), respectively. Therefore, the transition probability matrix between the V "germline" region and V-D "junction" region has 1 row and 9 columns.</p>
<ul>
<li>The first matrix entry \(P(V \rightarrow (V, 3))\) is equal to the <code>partis</code>-inferred transition probability going from position 2 to position 3 in germline gene \(\{V\}\).</li>
<li>The second entry \(P(V \rightarrow (V, 4))\) is equal to 0 because it is impossible to skip over positions in any germline gene.</li>
<li>Similarly, \(P(V \rightarrow (D, 1))\) is equal to \(P(V \rightarrow (V, \text{end})) P(D) P((D, \text{init}) \rightarrow (D, 1))\).</li>
</ul>
<p>Note that the \(\text{init}\) and \(\text{end}\) states in a germline gene represent the initial and ending states used in <code>partis</code> HMM germline parameter files. The other matrix entries, the ( \(9 \times 9\)) V-D "junction" transition probability matrix, and the ( \(9 \times 1\)) [V-D "junction"]-to-[D "germline"] transition probability matrix are filled using analogous logic. However, when computing the transition probabilities between the V-D "junction" region and D "germline" region, we must account for the transitions within the D "germline" region as well.</p>
<h1><a class="anchor" id="emission_prob"></a>
HMM hidden state emission probability matrices</h1>
<p>In each "germline" region, we have an emission probability row vector that stores an entry for every "germline" state in that region. The S-W example alignment shown in <a class="el" href="index.html#state_space">HMM hidden state space</a> has a V "germline" state space equal to \(\{V_{01}, V_{99}\}\) and thus the corresponding emission probability row vector has 2 entries. Because a "germline" region spans at least one site position in the alignment, each "germline" emission probability is defined as the product over all site-wise emission probabilities in the region. In the aforementioned example, the "germline" emission probabilities for the \(V_{01}\) and \(V_{99}\) states are products of the site-specific emission probabilities from position 0 to position 3.</p>
<p>In each "junction" region, we use an emission probability matrix with entries for each site position and "junction" state in that region. For the same example used above, the V-D "junction" emission probability matrix has 2 rows and 15 columns. Some "junction" states do not occur at particular site positions (i.e. \((D_{01}, 0)\) at position 4) and these cases are assigned emission probabilities of 0.</p>
<h1><a class="anchor" id="xmsa"></a>
Phylo-HMM "expanded" multiple sequence alignment (xMSA)</h1>
<p>In phylo-HMMs, per-column phylogenetic likelihoods take the place of emission probabilities in classical HMMs. Here, we describe how the notion of an "expanded" multiple sequence alignment (xMSA) makes phylo-HMM computation efficient. Suppose we have an S-W alignment as shown in <a class="el" href="index.html#sw_alignment">Smith-Waterman alignment information</a>, but instead of the single input sequence <code>ACAGTACCCTGTTNN</code>, we have a clonal family with 3 sequences (see diagram below).</p>
<div class="image">
<img src="msa.jpg" alt=""/>
</div>
<p>The phylogenetic likelihoods depend on the observed bases at each site and the associated hidden naive bases as well. We map each MSA site position and possible hidden naive base to an xMSA site position and compute the phylogenetic log-likelihood at each xMSA site only once. The important takeaway is that even though many pairs of MSA site positions and hidden naive bases occur more than once, we never have to compute the corresponding phylogenetic log-likelihoods more than once. The xMSA for the aforementioned example is displayed in the graphic below, where the top row of the alignment cycles through all of the possible bases of the "expanded" naive sequence.</p>
<div class="image">
<img src="xmsa.jpg" alt=""/>
</div>
<p>In this example,</p>
<ul>
<li>The first 4 columns of the MSA (<code>TCC</code>, <code>AAG</code>, <code>ACT</code>, <code>AAA</code>) are guaranteed to match with the first 4 bases of \(V\) (<code>NATG</code>) so there is no MSA expansion.</li>
<li>The next 4 columns are in the V-D "junction" region so we do not know the naive base with certainty. Thus, we repeat these columns with all possible naive bases as arbitrary non-templated insertions could fill the entire region.</li>
<li>The next MSA column (<code>CCG</code>) is in the D "germline" region and is guaranteed to match with the germline base <code>A</code>.</li>
</ul>
<p>Note that the V-D "junction" states \((D, 0)\) and \((D, N_G)\) at MSA site position 5 both map to the same xMSA site position (13) because they represent the same likelihood calculation.</p>
<h1><a class="anchor" id="scaling"></a>
HMM scaling for numeric underflow</h1>
<p>Given that there can be a potentially large number of sequences in a clonal family, we must account for possible numerical underflow in the forward algorithm. We define a large numeric constant <code>SCALE_FACTOR</code> and check that the computed forward probabilities are all above <code>1.0 / SCALE_FACTOR</code>. If not, we multiply the forward probabilities by <code>SCALE_FACTOR</code> enough times until no forward probability is less than <code>1.0 / SCALE_FACTOR</code> and record the number of times we multiplied by <code>SCALE_FACTOR</code>. We keep a running total of these scaler counts as the forward algorithm progresses along the site positions of the HMM and undo the scaling according to the final scaler count to compute the HMM log-likelihood. </p>
</div></div><!-- PageDoc -->
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Mon Mar 16 2020 16:55:42 for linearham by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.17
</small></address>
</body>
</html>
